source/core/lethe_grid_tools.cc

```diff
@@ -1367,3 +1367,194 @@ template std::
     const std::vector<Particles::ParticleIterator<3>> &particles,
     const unsigned int &n_particles_in_base_cell);
+
+
+template <int dim>
+double
+LetheGridTools::find_point_triangle_distance(
+  const std::vector<Point<dim>> &triangle,
+  const Point<dim> &             pos)
+{
+  auto &p_0 = triangle[0];
+  auto &p_1 = triangle[1];
+  auto &p_2 = triangle[2];
+
+  const Tensor<1, dim> e_0 = p_1 - p_0;
+  const Tensor<1, dim> e_1 = p_2 - p_0;
+
+  const Tensor<1, dim> normal      = cross_product_3d(e_0, e_1);
+  const double         norm_normal = normal.norm();
+  Tensor<1, dim>       unit_normal = normal / norm_normal;
+  Tensor<1, 3>         unit_normal_3d;
+  Point<3>             pt_in_triangle_3d;
+
+  const double a   = e_0.norm_square();
+  const double b   = scalar_product(e_0, e_1);
+  const double c   = e_1.norm_square();
+  const double det = a * c - b * b;
+
+
+  Tensor<1, dim> vector_to_plane;
+  Point<dim>     pt_in_triangle;
+
+  vector_to_plane = p_0 - pos;
+
+  // Do the full calculation taken from Eberly 2003
+  const double d = scalar_product(e_0, vector_to_plane);
+  const double e = scalar_product(e_1, vector_to_plane);
+
+  // Calculate necessary values;
+  double s = b * e - c * d;
+  double t = b * d - a * e;
+
+  if (s + t <= det)
+    {
+      if (s < 0)
+        {
+          if (t < 0)
+            {
+              // Region 4
+              if (d < 0)
+                {
+                  t = 0;
+                  if (-d >= a)
+                    s = 1;
+                  else
+                    s = -d / a;
+                }
+              else
+                {
+                  s = 0;
+                  if (e >= 0)
+                    t = 0;
+                  else if (-e >= c)
+                    t = 1;
+                  else
+                    t = e / c;
+                }
+            }
+          else
+            {
+              // Region 3
+              s = 0;
+              if (e >= 0)
+                t = 0;
+              else if (-e >= c)
+                t = 1;
+              else
+                t = -e / c;
+            }
+        }
+      else if (t < 0)
+        {
+          // Region 5
+          t = 0;
+          if (d >= 0)
+            s = 0;
+          else if (-d >= a)
+            s = 1;
+          else
+            s = -d / a;
+        }
+      else
+        {
+          // Region 0
+          const double inv_det = 1. / det;
+          s *= inv_det;
+          t *= inv_det;
+
+          // In region 0, normal vector is the face normal vector
+          // Cast unit_normal to a tensor<1, 3>
+          if constexpr (dim == 3)
+            unit_normal_3d = unit_normal;
+
+          if constexpr (dim == 2)
+            unit_normal_3d = tensor_nd_to_3d(unit_normal);
+        }
+    }
+  else
+    {
+      if (s < 0)
+        {
+          // Region 2
+          const double tmp0 = b + d;
+          const double tmp1 = c + e;
+          if (tmp1 > tmp0)
```

Comment:
I know it's a math thing and it doesn't really represent anything obvious, but some short explanation here and there would help :)

Reply:
Yeah, I'll just add a more complete reference to the original document, it's kinda hard to explain this without rewriting the entire reference.

Reply:
Ok!